src/core/server/plugins/plugin.ts

```diff
@@ -136,14 +136,14 @@ export interface DiscoveredPluginInternal extends DiscoveredPlugin {
  * @public
  */
 export interface Plugin<
-  TSetup,
-  TStart,
-  TPluginsSetup extends Record<PluginName, unknown> = {},
-  TPluginsStart extends Record<PluginName, unknown> = {}
```

Comment:
Why this change? `Record<PluginName, unknown>` is more explicit that the keys will be strings and forces you to construct your own time to deal with `unknown`.

Reply:
> ... explicit that the keys will be strings ...

Exactly, but `Record` does not just say that keys *can be strings*, but that *any string is a valid key*. Whereas if you extend `{}` you must specify the exact valid properties (not strings).

![image](https://user-images.githubusercontent.com/9773803/60203133-4ab7b280-984c-11e9-8f8c-9180c7a31497.png)

So basically in the screenshot above you can see TypeScript if fine giving "non-existing" `a.foo` to some JavaScript function (which I believe should be an error).

Reply:
Also, there is no `PluginName` alias in `./public`.

Is `PlaginName` always intended to just be an alias for `string`?

```ts
/**
 * Dedicated type for plugin name/id that is supposed to make Map/Set/Arrays
 * that use it as a key or value more obvious.
 *
 * @public
 */
export type PluginName = string;
```

If it was somehow possible to narrow it down.

```ts
export type PluginName = 'apm' | 'data' | 'canvas' | ... ;
```

Although there could be 3<sup>rd</sup> party plugins. And this would need to be auto-generated somehow.

Reply:
> Exactly, but `Record` does not just say that keys _can be strings_, but that _any string is a valid key_. Whereas if you extend `{}` you must specify the exact valid properties (not strings).
> 
> So basically in the screenshot above you can see TypeScript allowing to give "non-existing" `a.foo` to some JavaScript function (which I believe should be an error, if that `a` was a plugin contract). But it shows an error if you try the same with `b.foo`.

This is true but it's also true that you can't do anything useful with `unknown`. You can't pass it to a function or access any of it's properties without getting an error. The only thing this change does is remove information that the keys of the `plugins` argument will be `PluginName`. I prefer that this information be retained.

> Is PluginName always intended to just be an alias for string?
> If it was somehow possible to narrow it down.

We can't do this currently because we can't mix x-pack and OSS code. Maybe we could find a way to generate this, but I don't think it's particularly useful right now.

One way we could make this more useful is with an interface of default types that can be re-opened / merged by plugins.

```ts
interface PluginsStart {
  [pluginName: PluginName]: unknown;
}

// Each plugin can "reopen" this interface
declare module '../../src/core/public' {
  interface PluginsStart {
    timepicker?: TimepickerStart;
  }
}

// Dependent plugins could then declare which dependencies are available
// This would have to be manually kept in sync with the plugin's kibana.json
type RequiredDeps = 'data' | 'timepicker';
type OptionalDeps = 'telemetry';
type MyPluginsStart =
  Required<Pick<PluginsStart, RequiredDeps>> &
  Partial<Pick<PluginsStart, OptionalDeps>>;
// => { data: DataStart, timepicker: TimepickerStart, telemetry?: TelemetryStart }

Reply:
We could even provide a generic type for generating this:

```ts
type AvailablePluginsStart<Req extends keyof PluginsStart = never, Opt extends keyof PluginsStart = never>
  = Required<Pick<PluginsStart, Req>> & Partial<Pick<PluginsStart, Opt>>;

type MyPluginsStart = AvailablePluginsStart<'data' | 'timepicker', 'telemetry'>;
```

Reply:
> [..] You can't pass it to a function or access any of it's properties without getting an error. [..]

I was trying to show in my example above that you can pass it to a function.

Here is another concrete example from Core's source code. If you use `Record`, you can pass anything to a function, because `Record<string, ...>` means *any property is valid*:

![image](https://user-images.githubusercontent.com/9773803/60335499-659a3c00-999e-11e9-9d17-47a04bb2c7f1.png)

Whereas, if you restrict the `TPluginSetup` type, you get an error when you want to access property that does not exist:

![image](https://user-images.githubusercontent.com/9773803/60335682-c88bd300-999e-11e9-89a5-e4ad32967c20.png)

Note, this change still allows the rest of code work as before. Is there anything you cannot achieve with `{}` which you can with `Recrod<string, ...>`?

Reply:
>  If you use Record, you can pass anything to a function, because Record<string, ...> means any property is valid:

Not really. that works due to lax of type signature `alert(message?: any)`. A function with strict arguments will have TS error.
![2019-07-01_15-52-45](https://user-images.githubusercontent.com/3198181/60441932-49a1cf00-9c18-11e9-8f32-42cfb0bb452c.png)

I don't see any harm to add your changes to `Plugin` type definitions. Could you update other related places to unify their signatures? 
[1](https://github.com/restrry/kibana/blob/62b0115c85457aba4f257e13b52bbc2d6fef84d9/src/core/public/plugins/plugin_context.ts)
[2](https://github.com/restrry/kibana/blob/62b0115c85457aba4f257e13b52bbc2d6fef84d9/src/core/public/plugins/plugin.ts)
[3](https://github.com/restrry/kibana/blob/62b0115c85457aba4f257e13b52bbc2d6fef84d9/src/core/server/plugins/plugin.ts)

Reply:
> Not really. that works due to lax of type signature `alert(message?: any)`.

That's the whole point, you rely on the fact that the *other function* will "type check" the property. But `plugins` object itself allows to access any property. If the other function, for example, is a JavaScript function without a type signature, it happily accepts that property.

> Could you update other related files to unify `Setup/Start` signatures?

Do you mean all `TPluginsSetup` and `TPluginsStart` types in those files?

Reply:
> Yes. Do you mean all TPluginsSetup and TPluginsStart types in those files?

correct